src/kernels/kernelConnector.ts

```diff
@@ -0,0 +1,412 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+'use strict';
+
+import {
+    IKernel,
+    KernelConnectionMetadata,
+    IKernelProvider,
+    isLocalConnection,
+    KernelInterpreterDependencyResponse
+} from './types';
+import { Memento, NotebookDocument, NotebookController } from 'vscode';
+import { ICommandManager, IVSCodeNotebook, IApplicationShell } from '../platform/common/application/types';
+import { traceVerbose, traceWarning } from '../platform/logging';
+import { Resource, IMemento, GLOBAL_MEMENTO, IDisplayOptions, IDisposable } from '../platform/common/types';
+import { createDeferred, createDeferredFromPromise, Deferred } from '../platform/common/utils/async';
+import { DataScience } from '../platform/common/utils/localize';
+import { sendKernelTelemetryEvent } from '../telemetry/telemetry';
+import { IServiceContainer } from '../platform/ioc/types';
+import { Telemetry, Commands } from '../webviews/webview-side/common/constants';
+import { clearInstalledIntoInterpreterMemento } from './installer/productInstaller';
+import { Product } from './installer/types';
+import { INotebookControllerManager } from '../notebooks/types';
+import { findNotebookEditor, selectKernel } from '../notebooks/controllers/kernelSelector';
+import { KernelDeadError } from '../platform/errors/kernelDeadError';
+import { noop } from '../platform/common/utils/misc';
+import { IInteractiveWindowProvider } from '../interactive-window/types';
+import { IDataScienceErrorHandler } from '../platform/errors/types';
+import { IStatusProvider } from '../platform/progress/types';
+import { IRawNotebookProvider } from './raw/types';
+import { IVSCodeNotebookController } from '../notebooks/controllers/types';
+import { getDisplayNameOrNameOfKernelConnection } from './helpers';
+
+/**
+ * Class used for connecting a controller to an instance of an IKernel
+ */
+export class KernelConnector {
```

Comment:
This is all the functions that used to be in helper.ts for connecting to a kernel. There was even a little bit of state. Figured it made sense to put the functions into a class to hold the state.